### PR TITLE
Upgrade to gradle 5 and lots of other improvements

### PR DESCRIPTION
**Pull request by smac89**
- Remove complex logic of choosing path to `flatc` executable. Either use what the user supplied or assume compiler is on path
- Use the builtin `Delete` task rather than rolling our own
- Use callbacks to determine when plugins such as `java` and `idea` are applied and then mixin our own logic
- Added a default dependency on flatbuffers-java which can easily be overridden.

### CODE REVIEW

**Comment by gregwhitaker** [comment]
@smac89 any reason this was closed? was just getting ready to review

**Comment by smac89** [comment]
@gregwhitaker I will review the PR and reopen if I don't find anything wrong with it. I will try to do that by the end of this week.

**Comment by smac89** [comment]
@gregwhitaker This PR is ready for review :smile:

File: ci/travis.sh [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env bash
 # This script builds the project in Travis
+export GRADLE_OPTS="-Dorg.gradle.daemon=false"
```
[the remark this answers is not in the document]
Comment (smac89) [reply ci/travis.sh]: I've been bitten before and that's why I take this precaution on gradle builds using travis.

**Pull request merged**